pkg/blocks/step.go

```diff
@@ -22,241 +22,176 @@ package blocks
 import (
 	"errors"
 	"fmt"
-	"runtime"
-	"strings"
 
 	"github.com/facebookincubator/ttpforge/pkg/logging"
-	"github.com/facebookincubator/ttpforge/pkg/outputs"
-	"go.uber.org/zap"
 	"gopkg.in/yaml.v3"
 )
 
-// Constants representing supported executor types.
-const (
-	ExecutorPython     = "python3"
-	ExecutorBash       = "bash"
-	ExecutorSh         = "sh"
-	ExecutorPowershell = "powershell"
-	ExecutorRuby       = "ruby"
-	ExecutorBinary     = "binary"
-	ExecutorCmd        = "cmd.exe"
-)
-
-// StepType denotes the type of a step in a TTP.
-type StepType string
-
-// Constants for defining the types of steps available.
-const (
-	StepCreateFile = "createFileStep"
-	StepUnset      = "unsetStep"
-	StepFile       = "fileStep"
-	StepFetchURI   = "fetchURIStep"
-	StepBasic      = "basicStep"
-	StepSubTTP     = "subTTPStep"
-	StepCleanup    = "cleanupStep"
-	StepEdit       = "editStep"
-)
+type CommonStepFields struct {
+	Name        string `yaml:"name,omitempty"`
+	Description string `yaml:"description,omitempty"`
 
-// Act represents a single action within a TTP (Tactics, Techniques,
-// and Procedures) step.
-//
-// Condition: The condition that needs to be satisfied for the Act to execute.
-// Environment: Environment variables used during the Act's execution.
-// Name: The unique name of the Act.
-// WorkDir: The working directory of the Act.
-// Type: The type of the Act (e.g., Command, File, or Setup).
-// success: Indicates whether the execution of the Act was successful.
-// stepRef: Reference to other steps in the sequence.
-// output: The output of the Act's execution.
-type Act struct {
-	Condition   string                  `yaml:"if,omitempty"`
-	Environment map[string]string       `yaml:"env,omitempty"`
-	Name        string                  `yaml:"name"`
-	WorkDir     string                  `yaml:"-"`
-	Outputs     map[string]outputs.Spec `yaml:"outputs,omitempty"`
-	Type        StepType                `yaml:"-"`
+	// CleanupSpec is exported so that UnmarshalYAML
+	// can see it - however, it should be considered
+	// to be a private detail of this file
+	// and not referenced elsewhere in the codebase
+	CleanupSpec yaml.Node `yaml:"cleanup,omitempty"`
 }
 
-// CleanupAct interface is implemented by anything that requires a cleanup step.
-type CleanupAct interface {
-	Cleanup(execCtx TTPExecutionContext) (*ActResult, error)
-	StepName() string
-	SetDir(dir string)
-	IsNil() bool
-	Validate(execCtx TTPExecutionContext) error
+// Step contains a TTPForge executable action
+// and its associated cleanup action (if specified)
+type Step struct {
+	CommonStepFields
```

Comment:
All of the nasty UnmarshalYAML logic that previously had to be duplicated to every step type now lives in this single file, and has been further simplified